barretenberg/cpp/src/barretenberg/stdlib/primitives/group/cycle_group.cpp

```diff
@@ -550,158 +545,149 @@ cycle_group<Builder> cycle_group<Builder>::checked_unconditional_subtract(const
-    const bool_t double_predicate = (x_coordinates_match && !y_coordinates_match).normalize();
-    const bool_t infinity_predicate = (x_coordinates_match && y_coordinates_match).normalize();
-    if (!infinity_predicate.is_constant()) {
-        infinity_predicate.get_context()->update_used_witnesses(infinity_predicate.get_normalized_witness_index());
-    }
+
     const field_t x1 = x;
     const field_t y1 = y;
     const field_t x2 = other.x;
     const field_t y2 = other.y;
+
+    // Execute point addition with modified lambda = (-y2 - y1)/(x2 - x1 + x_coordinates_match) to avoid the possibility
+    // of division by zero.
     const field_t x_diff = x2.add_two(-x1, x_coordinates_match);
-    // Computes lambda = (-y2-y1)/x_diff, using the fact that x_diff is never 0
+    // Compute lambda in one of two ways depending on whether either numerator or denominator is constant or not
     field_t lambda;
     if ((y1.is_constant() && y2.is_constant()) || x_diff.is_constant()) {
         lambda = (-y2 - y1).divide_no_zero_check(x_diff);
     } else {
-        lambda =
-            field_t::from_witness(this->get_context(other), (-y2.get_value() - y1.get_value()) / x_diff.get_value());
+        Builder* context = get_context(other);
+        lambda = field_t::from_witness(context, (-y2.get_value() - y1.get_value()) / x_diff.get_value());
         // We need to manually propagate the origin tag
         lambda.set_origin_tag(OriginTag(x_diff.get_origin_tag(), y1.get_origin_tag(), y2.get_origin_tag()));
+        // Constrain x_diff * lambda = -y2 - y1
         field_t::evaluate_polynomial_identity(x_diff, lambda, y2, y1);
     }
+    const field_t x3 = lambda.madd(lambda, -(x2 + x1)); // x3 = lambda^2 - x1 - x2
+    const field_t y3 = lambda.madd(x1 - x3, -y1);       // y3 = lambda * (x1 - x3) - y1
+    cycle_group sub_result(x3, y3, /*is_infinity=*/x_coordinates_match);
 
-    const field_t x3 = lambda.madd(lambda, -(x2 + x1));
-    const field_t y3 = lambda.madd(x1 - x3, -y1);
-    cycle_group add_result(x3, y3, x_coordinates_match);
-
+    // Compute the doubling result
     const cycle_group dbl_result = dbl();
 
-    // dbl if x_match, !y_match
-    // infinity if x_match, y_match
-    field_t result_x = field_t::conditional_assign(double_predicate, dbl_result.x, add_result.x);
-    field_t result_y = field_t::conditional_assign(double_predicate, dbl_result.y, add_result.y);
-
-    if constexpr (IsUltraBuilder<Builder>) {
-        if (result_x.get_context()) {
-            result_x.get_context()->update_used_witnesses(result_x.witness_index);
-        }
-        if (result_y.get_context()) {
-            result_y.get_context()->update_used_witnesses(result_y.witness_index);
-        }
-    }
+    // If the subtraction amounts to a doubling then the result is the doubling result, else the subtraction result.
+    const bool_t double_predicate = (x_coordinates_match && !y_coordinates_match);
```

Comment:
I wonder if we should have a stricter check for `double_predicate` here, something like:
```cpp
const bool_t y_coordinates_cancel = (y + other.y).is_zero();
const bool_t double_predicate = (x_coordinates_match && y_coordinates_cancel);
```

Reply:
isn't this equivalent to what we have now? If the x-coords agree and the y coords don't then it must be that $(x_1, y_1) == (x_2, -y_2)$. Maybe it would be more clear to have the condition you suggest though